### DISCUSSION TITLE
Too many attributes error

### DISCUSSION THREAD

**Discussion by ricardoekm** [visgl/deck.gl#8096] [Q&A] [answered]
Hi,
I'm trying to write a custom layer that extends the PathLayer. I've the [data filter extension](https://deck.gl/docs/api-reference/extensions/data-filter-extension) enabled.

If I add two attributes (isAnimated and timestamps) in the attribute manager I get the following error:

```
index.js:1 deck: initialization of LineLayer({id: '#pipes#line'}): Error linking: Too many attributes (instancePickingColors)
Error: Error linking: Too many attributes (instancePickingColors)

at Program._compileAndLink (program.js:344:1)
```

It seems an error from Angle, saying I've reached the limit of vertex attributes. Inspecting the attribute manager and calling the getShaders method, I get the following result:

<img width="787" alt="Screenshot 2023-09-04 at 16 50 46" src="https://github.com/visgl/deck.gl/assets/747894/e7733255-d5b5-44d7-ba27-b98fe2629618">

Is there anything I can do to fix it? Thanks.

**Comment by ibgreen** [comment, marked as answer]
Most GPUs only support 16 vertex attributes. It is a hard limit, outside of deck.gl's control. Each attribute can be a vec4 (Float32x4), so the main technique you can use is to share data from two "logical" attributes in a single attribute (e.g. using attr.x attr.y for something and attr.z and attr.w for something else)

**Reply by ricardoekm** [reply]
Thanks @ibgreen !!
